src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs

```diff
@@ -331,108 +331,142 @@ public struct S
-                // (14,9): warning CS8655: Call to non-readonly member 'Equals' from a 'readonly' member results in an implicit copy of 'this'.
-                //         Equals(null);
-                Diagnostic(ErrorCode.WRN_ImplicitCopyInReadOnlyMember, "Equals").WithArguments("Equals", "this").WithLocation(14, 9));
+            var method = comp.GetMember<NamedTypeSymbol>("S").GetMethod("M");
+            // Symbol APIs always return the declaration part of the partial method.
+            Assert.False(method.IsDeclaredReadOnly);
+            Assert.False(method.IsEffectivelyReadOnly);
         }
 
         [Fact]
-        public void ReadOnlyMethod_Override_AssignThis()
+        public void ReadOnlyMethod_Partial_03()
         {
             var csharp = @"
-public struct S
+public partial struct S
 {
     public int i;
+    readonly partial void M();
+}
 
-    // error
-    public readonly override string ToString() => (i++).ToString();
-    public readonly override int GetHashCode() => (i++).GetHashCode();
-    public readonly override bool Equals(object o) => (i++).Equals(o);
+public partial struct S
+{
+    partial void M()
+    {
+        i++;
+    }
 }
 ";
             var comp = CreateCompilation(csharp);
             comp.VerifyDiagnostics(
-                // (7,52): error CS1604: Cannot assign to 'i' because it is read-only
-                //     public readonly override string ToString() => (i++).ToString();
-                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "i").WithArguments("i").WithLocation(7, 52),
-                // (8,52): error CS1604: Cannot assign to 'i' because it is read-only
-                //     public readonly override int GetHashCode() => (i++).GetHashCode();
-                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "i").WithArguments("i").WithLocation(8, 52),
-                // (9,56): error CS1604: Cannot assign to 'i' because it is read-only
-                //     public readonly override bool Equals(object o) => (i++).Equals(o);
-                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "i").WithArguments("i").WithLocation(9, 56));
+                // (10,18): error CS8662: Both partial method declarations must be readonly or neither may be readonly
+                //     partial void M()
+                Diagnostic(ErrorCode.ERR_PartialMethodReadOnlyDifference, "M").WithLocation(10, 18));
+
+            var method = comp.GetMember<NamedTypeSymbol>("S").GetMethod("M");
+            // Symbol APIs always return the declaration part of the partial method.
+            Assert.True(method.IsDeclaredReadOnly);
+            Assert.True(method.IsEffectivelyReadOnly);
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_ExplicitInterfaceImplementation()
+        {
+            var csharp = @"
+public interface I
+{
+    void M();
+}
+
+public struct S : I
+{
+    readonly void I.M()
```

Comment:
>readonly void I.M() [](start = 4, length = 19)

nit: If I understand properly, the only benefit of adding `readonly` here is to the author of the method (the compiler will help preserve intended semantics), but not the consumer of the method.
If that's correct, consider adding a call in this body that produces a diagnostic.

Reply:
That's true. Will do.